### PR TITLE
Add ECC support and solve wrong Length status codes with SM card

### LINKED ISSUE #1073: ePass2003 EC not supported
It seems that OpenSC (**0.16.0** Arch Linux build) don't support elliptic curves on Feitian Technologies, Inc. ePass2003 USB token (**096e:0807**).

    ~> opensc-tool --list-algorithms
    Using reader with a card: Feitian ePass2003 00 00
    Algorithm: rsa
    Key length: 512
    Flags: onboard key generation padding ( none ) hashes ( )

    Algorithm: rsa
    Key length: 768
    Flags: onboard key generation padding ( none ) hashes ( )

    Algorithm: rsa
    Key length: 1024
    Flags: onboard key generation padding ( none ) hashes ( )

    Algorithm: rsa
    Key length: 2048
    Flags: onboard key generation padding ( none ) hashes ( )

    ~> pkcs11-tool -k --key-type EC:prime256v1
    Using slot 0 with a present token (0x0)
    error: Generate EC key mechanism not supported

    ~> pkcs15-init --generate-key EC:prime256v1 --key-usage sign --auth-id 1
    Using reader with a card: Feitian ePass2003 00 00
    Failed to generate key: Not supported

Bay the way: why is epass2003.profile described as `pkcs15 profile for entersafe`

### LINKED ISSUE #1115: feitian epass2003 is not working with opensc 0.17.0 on Fedora 26
```
opensc-tool --list-readers                                                                                                                               
# Detected readers (pcsc)
Nr.  Card  Features  Name
0    Yes             FT ePass2003Auto 00 00
```

pkcs15-init --erase-card -vvv  **(this works fine)**
pkcs15-init --create-pkcs15 --profile pkcs15+onepin --label "sandys" -vvv **(this works fine)**
pkcs15-init --auth-id 1 --generate-key rsa/2048 --key-usage sign,decrypt -vvv **(this fails)**

> Using reader with a card: FT ePass2003Auto 00 00
> PKCS#15 binding failed: Unsupported card
> 

the detailed logs have been attached. grep for "sss>" to see the different commands that i have run separately.
[debug.log.zip](https://github.com/OpenSC/OpenSC/files/1191090/debug.log.zip)

> 
> uname -a                                                                                                                                                 
> Linux menzoberranzan 4.11.11-300.fc26.x86_64 #1 SMP Mon Jul 17 16:32:11 UTC 2017 x86_64 x86_64 x86_64 GNU/Linux
> 
> LSB Version:	:core-4.1-amd64:core-4.1-noarch:cxx-4.1-amd64:cxx-4.1-noarch:desktop-4.1-amd64:desktop-4.1-noarch:languages-4.1-amd64:languages-4.1-noarch:printing-4.1-amd64:printing-4.1-noarch
> Distributor ID:	Fedora
> Description:	Fedora release 26 (Twenty Six)
> Release:	26
> Codename:	TwentySix

> 
> Installed Packages
> Name         : opensc
> Version      : 0.17.0
> Release      : 1.fc26
> Arch         : x86_64
> Size         : 3.3 M
> Source       : opensc-0.17.0-1.fc26.src.rpm

> Name         : pcsc-tools
> Version      : 1.4.25
> Release      : 2.fc26
> Arch         : x86_64
> Size         : 836 k
> Source       : pcsc-tools-1.4.25-2.fc26.src.rpm
> 
> Name         : pcsc-lite-ccid
> Version      : 1.4.27
> Release      : 1.fc26
> Arch         : x86_64
> Size         : 1.4 M

### PR DESCRIPTION
**Pull request by FeitianSmartcardReader**
Fixes #1073
Fixes #1115

- [ ] Documentation is added or updated
- [ ] New files have a LGPL 2.1 license statement
- [ ] Tested with the following card: 
	- [x] tested PKCS#11
	- [ ] tested Windows Minidriver
	- [x] tested macOS Tokend

### CODE REVIEW

**Comment by FeitianSmartcardReader** [comment]
@frankmorgner could you please find time have a check my PR and confirm? some of our customer waiting now, thanks

**Comment by frankmorgner** [comment]
I'll be out of office for quite some time so I doubt that I have a lot of time reviewing big changes. If you have time constraints, please consider managing and distributing your own copy of OpenSC to your customers. It will surely take some time for a new release of OpenSC.

**Comment by majkrzak** [comment]
You disappointed me, @frankmorgner. You had time to merge #1144 without even looking at code (It is quite impossible for me that someone could accept such thing).

**Comment by martinpaljak** [comment]
@majkrzak Could you elaborate?

File: src/libopensc/card-epass2003.c [hunk 1]
```diff
@@ -2129,19 +2235,36 @@ internal_write_rsa_key(struct sc_card *card, unsigned short fid, struct sc_pkcs1
...
-	sha1_digest(data, datalen, data_hash);
+	if(mechanismType | SC_ALGORITHM_ECDSA_HASH_SHA1)
+	{
+		unsigned char data_hash[24] = { 0 };
+		size_t len = 0;
 
-	len = REVERSE_ORDER4(datalen);
-	memcpy(&data_hash[20], &len, 4);
-	memcpy(hash, data_hash, 24);
+		sha1_digest(data, datalen, data_hash);
```
Comment (frankmorgner) [inline src/libopensc/card-epass2003.c]: Do you want to do error checking?

File: src/libopensc/card-epass2003.c [hunk 2]
```diff
@@ -1640,7 +1684,55 @@ static int epass2003_decipher(struct sc_card *card, const u8 * data, size_t data
...
+		return SC_ERROR_INVALID_ARGUMENTS;
+	
+	exdata = (epass2003_exdata *)card->drv_data;
 
+	if(exdata->currAlg == SC_ALGORITHM_EC)
+	{
+		unsigned char hash[HASH_LEN] = { 0 };
+		if(exdata->ecAlgFlags | SC_ALGORITHM_ECDSA_HASH_SHA1)
+		{
+			hash_data(data, datalen, hash, SC_ALGORITHM_ECDSA_HASH_SHA1);
```
Comment (frankmorgner) [inline src/libopensc/card-epass2003.c]: Do you want to do error checking, which is done in the rest of the code for `hash_data()`?

File: src/libopensc/card-epass2003.c [hunk 3]
```diff
@@ -1640,7 +1684,55 @@ static int epass2003_decipher(struct sc_card *card, const u8 * data, size_t data
...
+			sc_format_apdu(card, &apdu, SC_APDU_CASE_3,0x2A, 0x9E, 0x9A);
+			memset(sbuf, 0, sizeof(sbuf));
+			memcpy(sbuf, hash, 0x14);
+			apdu.data = sbuf;
+			apdu.lc = 0x14;
+			apdu.datalen = 0x14;
+		}
+		else if (exdata->ecAlgFlags | SC_ALGORITHM_ECDSA_HASH_SHA256)
+		{
+			hash_data(data, datalen, hash, SC_ALGORITHM_ECDSA_HASH_SHA256);
```
Comment (frankmorgner) [inline src/libopensc/card-epass2003.c]: Do you want to do error checking?

File: src/libopensc/card-epass2003.c [hunk 4]
```diff
@@ -2129,19 +2235,36 @@ internal_write_rsa_key(struct sc_card *card, unsigned short fid, struct sc_pkcs1
...
+		len = REVERSE_ORDER4(datalen);
+		memcpy(&data_hash[20], &len, 4);
+		memcpy(hash, data_hash, 24);
+	}
+	else if(mechanismType | SC_ALGORITHM_ECDSA_HASH_SHA256)
+	{
+		unsigned char data_hash[36] = { 0 };
+		size_t len = 0;
+
+		sha256_digest(data, datalen, data_hash);
```
Comment (frankmorgner) [inline src/libopensc/card-epass2003.c]: Do you want to do error checking?

**Comment by frankmorgner** [comment]
@FeitianSmartcardReader your feedback is required: Do you plan to add error checking?

**Comment by FeitianSmartcardReader** [comment]
@frankmorgner thanks for the reminder, please help merge to master, some customer is waiting to using it.
And for error checking, not now, will do once got time. thanks for your help,

**Pull request merged**